### PR TITLE
grpc/service: Refactor middleware

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: Close https://github.com/tikv/pd/issues/8850

### What is changed and how does it work?

```commit-message
Organize the middleware code and handle them centrally.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/middleware.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	"google.golang.org/grpc"
+)
+
+type request interface {
+	GetHeader() *pdpb.RequestHeader
+}
+
+type forwardFn func(ctx context.Context, client *grpc.ClientConn, request any) (any, error)
+
+var forwardFns = map[string]forwardFn{
```
Comment (rleungx) [inline server/middleware.go:35]: I think it's more complicated.

Comment (okJiang) [reply server/middleware.go:35]: Isn't it easier to manage after the modifications compared to the previous definition scattered across each function? Could you elaborate on your viewpoint?

Comment (rleungx) [reply server/middleware.go:35]: There are too many hardcode things and you still need to add them one by one.

Comment (okJiang) [reply server/middleware.go:35]: From the current implementation, it seems that hardcoding cannot be avoided (it was hardcoded before as well). Based on the existing changes, where do you think it has become more complex?

Comment (rleungx) [reply server/middleware.go:35]: We don't need to care about the function name for now.

Comment (okJiang) [reply server/middleware.go:35]: If we need to maintain it, we have to pay attention to the function names, right? I think we should know which functions support a certain middleware and which do not.
> don't need to care about the function name

Or do you mean that all functions should support the forward middleware, so we don't need to worry about the function names? If that's the case, I don't think the impact is significant; placing a map here doesn't imply that we need to be overly concerned about function names, and it doesn't add much burden. If we didn't have this map, we would still need to define a `forwardFn` inside the function when adding new functions, which makes no difference.

If we still hold different views, perhaps we should listen to the opinions of other members. cc @JmPotato @nolouch @lhy1024

**Comment by okJiang** [comment]
I will fix the problem from ci after discussion

**Comment by okJiang** [comment]
Since there are no obvious benefits, this issue will be closed for now.

**Pull request closed without merging**